File: gh-data/src/arguments/parser.rs
use std::{
    env::args,
    io::{StdoutLock, Write, stdout},
    process::ExitCode,
    string::String,
    vec::Vec,
};

use super::{print_help::print_help_message, print_version::print_version_number};

use crate::github::{
    write_folliwng::following, write_repositories::repositories, write_stars::starred,
};

// Command Line Argument Tokenizer
pub fn tokenize_arguments() -> ExitCode {
    let command_line_arguments: Vec<String> = args().collect();
    let mut standard_output: StdoutLock = stdout().lock();

    if command_line_arguments.len() != 3 {
        writeln!(
            standard_output,
            "\x1b[31;1;3;4mFlag and User Required:\x1b[0m {:#?}",
            command_line_arguments
        )
        .unwrap();
        writeln!(standard_output, "").unwrap();
        print_help_message();
        eprintln!("\x1b[31;1;3;4mError(1) - Exiting GitHub Data Tool\x1b[0m");
        return ExitCode::FAILURE;
    } else {
        match command_line_arguments[1].trim() {
            "--following" | "--f" => {
                following(command_line_arguments[2].trim());
            }
            "help" | "--h" => {
                print_help_message();
            }
            "--repositories" | "--r" => {
                repositories(command_line_arguments[2].trim());
            }
            "--starred" | "--s" => {
                starred(command_line_arguments[2].trim());
            }

            "version" | "--v" => {
                print_version_number();
            }
            &_ => {
                writeln!(
                    standard_output,
                    "\x1b[31;1;3;4mUknown Command or Flag:\x1b[0m {:#?}",
                    command_line_arguments[1].trim()
                )
                .unwrap();
                writeln!(standard_output, "").unwrap();
                print_help_message();
                eprintln!("\x1b[31;1;3;4mError(1) - Exiting GitHub Data Tool\x1b[0m");
                return ExitCode::FAILURE;
            }
        };
    };

    return ExitCode::SUCCESS;
}

File: gh-data/src/arguments/parser.c
#include "./parser.h"

#include "../github/following.h"
#include "../github/repositories.h"
#include "../github/stars.h"
#include "./help_message.h"
#include "./version_message.h"

#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Parse Command Line Arguments
uint8_t parse_arguments(int length, char *arguments[]) {
  int index;

  if (length == 1 || length > 2) {
    printf("\x1b[31;1;3;4mFlag and User Required:\x1b[0m\n\n");
    for (index = 0; index < length; index = index + 1) {
      printf("%s\n\n", arguments[index]);
    }
    print_help();
    exit(EXIT_FAILURE);
  } else if (strcmp(arguments[1], "--following") == 0) {
    write_following();
  } else if (strcmp(arguments[1], "--f") == 0) {
    write_following();
  } else if (strcmp(arguments[1], "help") == 0) {
    print_help();
  } else if (strcmp(arguments[1], "--h") == 0) {
    print_help();
  } else if (strcmp(arguments[1], "--repositories") == 0) {
    write_repositories();
  } else if (strcmp(arguments[1], "--r") == 0) {
    write_repositories();
  } else if (strcmp(arguments[1], "--starred") == 0) {
    write_starred();
  } else if (strcmp(arguments[1], "--s") == 0) {
    write_starred();
  } else if (strcmp(arguments[1], "version") == 0) {
    print_version();
  } else if (strcmp(arguments[1], "--v") == 0) {
    print_version();
  } else {
    printf("\x1b[31;1;3;4mUknown Command or Flag:\x1b[0m\n\n");
    for (index = 0; index < length; index = index + 1) {
      printf("%s\n\n", arguments[index]);
    }
    print_help();
    exit(EXIT_FAILURE);
  }

  return 0;
}
